compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -120,6 +120,48 @@ pub fn overlapping_impls(
     Some(overlap(selcx, skip_leak_check, impl1_def_id, impl2_def_id, overlap_mode).unwrap())
 }
 
+/// Given an impl_def_id that "positively" implement a trait, check if the "negative" holds.
+pub fn negative_impl_holds(tcx: TyCtxt<'_>, impl_def_id: DefId, overlap_mode: OverlapMode) -> bool {
+    debug!("negative_impl_holds(impl1_header={:?}, overlap_mode={:?})", impl_def_id, overlap_mode);
+    // `for<T> (Vec<u32>, T): Trait`
+    let header = tcx.impl_trait_ref(impl_def_id).unwrap();
+
+    let infcx = tcx
+        .infer_ctxt()
+        .with_opaque_type_inference(DefiningAnchor::Bubble)
+        .intercrate(true)
+        .build();
+
+    // `[?t]`
+    let infer_substs = infcx.fresh_substs_for_item(DUMMY_SP, impl_def_id);
+
+    // `(Vec<u32>, ?t): Trait`
+    let trait_ref = header.subst(tcx, infer_substs);
+
+    // `(Vec<u32>, ?t): !Trait`
+    let trait_pred = tcx.mk_predicate(ty::Binder::dummy(ty::PredicateKind::Clause(
+        ty::Clause::Trait(ty::TraitPredicate {
+            trait_ref,
+            constness: ty::BoundConstness::NotConst,
+            polarity: ty::ImplPolarity::Negative,
+        }),
+    )));
+
+    // Ideally we would use param_env(impl_def_id) but that's unsound today.
+    let param_env = ty::ParamEnv::empty();
```

Comment:
To do this properly, what we ought to do is get the parameter environment and then apply the `infer_substs` above. The idea is that for some impl...

```rust
impl<A..Z> SomeTrait<T1..Tn> for T0
where WC
```

we want to test if...

* there exists types `A..Z` such that...
* assuming `WC` is true (ie., with the param env in scope)...
* `T0: !SomeTrait<T1..Tn>`

this would be an error, because that means that for at least *some* instance of this positive impl, there exists a negative impl.

Problem is: applying `infer_substs` to param-env will mean that we have inference variables in the environment, and I think our code doesn't like that.

It should be sound to not assume WC is true, because proving `A => X` is harder than proving `X` on its own, but I'm worried we may be *overly* conservative.

Reply:
inference variables in the param env generally cause issues, especially in the old solver. Don't remember exactly what breaks but iirc @BoxyUwU encountered a bunch of issues recently when she tried it.

I think a bigger issue is that the way `drop_candidate_in_favour_of` handles param env candidates is incomplete. This means that coherence in the old solver with trait goals in the param env is unsound because inference may incorrectly cause another nested goal to fail, resulting in an incorrect error. My current approach for the new solver is to disable this behavior if we're in coherence: #109724